src/ripple/protocol/impl/InnerObjectFormats.cpp

```diff
@@ -59,6 +59,24 @@ InnerObjectFormats::InnerObjectFormats()
             {sfNFTokenID, soeREQUIRED},
             {sfURI, soeOPTIONAL},
         });
+
+    add(sfVoteEntry.jsonName.c_str(),
+        sfVoteEntry.getCode(),
+        {
+            {sfAccount, soeREQUIRED},
+            {sfTradingFee, soeREQUIRED},
+            {sfVoteWeight, soeREQUIRED},
+        });
+
+    add(sfAuctionSlot.jsonName.c_str(),
+        sfAuctionSlot.getCode(),
+        {
+            {sfAccount, soeREQUIRED},
+            {sfExpiration, soeREQUIRED},
+            {sfDiscountedFee, soeREQUIRED},
```

Comment:
Because the current implementation always sets `sfDiscountedFee` to 0, make it `soeDEFAULT`. It will save a few bytes on disk, because it won't bother writing the 0, and will act like a 0 whenever it's read.

Reply:
If this is changed then the AMM unit-tests fail on `ledger_accept` in `Env::close()` and eventually there is a crash because of `LogicError()` call. I didn't dig more into this.

Reply:
Oops. You're right. The problem is that "default" fields don't like being explicitly set to the default value. The solution is to simply remove the line setting it to 0 in AMMCreate.cpp, and change where it's set in AMMBid.cpp to
```
        // sfDiscountedFee is a defaulted field, so remove it if the new fee is 0.
        if (fee == 0)
            auctionSlot.makeFieldAbsent(sfDiscountedFee);
        else
            auctionSlot.setFieldU32(sfDiscountedFee, fee);
```
It's a bit more complicated, but I think the space savings will make it worth it.

(I haven't gotten there yet, but I think doing the same thing for `sfTradingFee` would be worthwhile, since that can potentially be set to 0, too.)